### ISSUE TITLE
eat too much cpu on 3.0.5

### ISSUE THREAD

**Issue by fengmk2** [markdown-it/markdown-it#54]
Parse blew markdown, `3.0.5` use `0m20.459s`, and `3.0.4` use `0m2.981s`

```md
[@addyosmani](https://github.com/addyosmani),
[@borismus](https://github.com/borismus),
[@carsonmcdonald](https://github.com/carsonmcdonald),
[@chriseppstein](https://github.com/chriseppstein),
[@danwrong](https://github.com/danwrong),
[@davidmaxwaterman](https://github.com/davidmaxwaterman),
[@desandro](https://github.com/desandro),
[@hemanth](https://github.com/hemanth),
[@isaacs](https://github.com/isaacs),
[@josh](https://github.com/josh),
[@jrburke](https://github.com/jrburke),
[@marcelombc](https://github.com/marcelombc),
[@marcooliveira](https://github.com/marcooliveira),
[@mklabs](https://github.com/mklabs),
[@MrDHat](https://github.com/MrDHat),
[@necolas](https://github.com/necolas),
[@paulirish](https://github.com/paulirish),
[@richo](https://github.com/richo),
[@rvagg](https://github.com/rvagg),
[@sindresorhus](https://github.com/sindresorhus),
[@SlexAxton](https://github.com/SlexAxton),
[@sstephenson](https://github.com/sstephenson),
[@svnlto](https://github.com/svnlto),
[@tomdale](https://github.com/tomdale),
[@uzquiano](https://github.com/uzquiano),
```

repeat codes:

```js
var MarkdownIt = require('markdown-it');

    var md = new MarkdownIt({
      html: true,
      linkify: true,
    });
    var readme = fs.readFileSync(path.join(fixtures, 'eat-cpu.md'), 'utf8');
    md.render(readme);
```

**Comment by puzrin** [comment]
Problem is in this line change https://github.com/markdown-it/markdown-it/commit/8ca0b5b52b666847c972bf17cbd1df1710fd6689#diff-f4f9b75e4799aa5566d488b8f480de40L37

Change looks correct, but i don't see any broken tests on revert. Also, we should what happens with recursion on real references, because list can be long.

/cc @rlidwka , take a look please. I don't see check for `]:`

**Comment by puzrin** [comment]
https://github.com/markdown-it/markdown-it/blob/master/lib/parser_block.js#L18

`[reference]` seems to be a mistake. After playing with official dingus & digging spec, i didn't found a way to do such interrupt. This rule cause heavy recursion, trashing everything.

@fengmk2 if you can't wait at all - remove nested array from that line. Will publish fix a bit later, when @rlidwka appears online.

**Comment by rlidwka** [comment]
Fixed in refs branch.

It was an optimization attempt gone south. If we have a long list of references (like https://github.com/markdown-it/markdown-it/commit/533b5c89f40b7d8fa4ce6edc8e0a7550f25de90b), this parser would call `getLines` for each reference until the end of the paragraph. So first call would be `getLines(1, 50)`, second call would be `getLines(2, 50)`, etc. So I thought, it would be nice to optimize it a bit. It didn't go so well.

Anyway, the immediate solution would be to remove references from there.

And we have a benchmark that performs 3x times slower than CommonMark (in theory, their parser is linear on that input, and ours isn't), but it might be left for future discussions.

**Comment by fengmk2** [comment]
I use 3.0.4 right now.

**Comment by puzrin** [comment]
Fixed in published 3.0.6

**Comment by fengmk2** [comment]
cool! Thanks @puzrin @rlidwka
